In [1]:
import numpy as np
import pandas as pd

In [2]:
project_path = "/content/drive/MyDrive/Colab Notebooks/Hitachi/"

In [3]:
demand_forecast_data_1_path = "Data/Demand Forecasting/Demand Forecasting Demand Data upto Feb 21.csv"
demand_forecast_data_1 = pd.read_csv(project_path + demand_forecast_data_1_path)
demand_forecast_data_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27552 entries, 0 to 27551
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   datetime     27552 non-null  object 
 1   Demand (MW)  27552 non-null  float64
dtypes: float64(1), object(1)
memory usage: 430.6+ KB


In [4]:
demand_forecast_data_1.head()

datetime  Demand (MW)
0  1/1/2020 0:00        445.8
1  1/1/2020 1:00        424.5
2  1/1/2020 2:00        423.5
3  1/1/2020 3:00        418.8
4  1/1/2020 4:00        414.8

In [5]:
demand_forecast_data_1.tail()

datetime  Demand (MW)
27547  2/21/2023 19:00        498.7
27548  2/21/2023 20:00        498.8
27549  2/21/2023 21:00        506.4
27550  2/21/2023 22:00        478.8
27551  2/21/2023 23:00        457.7

In [6]:
demand_forecast_data_2_path = "Data/Demand Forecasting/Demand Forecasting Weather Data upto Feb 28.csv"
demand_forecast_data_2 = pd.read_csv(project_path + demand_forecast_data_2_path)
demand_forecast_data_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27720 entries, 0 to 27719
Data columns (total 26 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Name              27720 non-null  object 
 1   datetime          27720 non-null  object 
 2   Temperature       27720 non-null  float64
 3   feelslike         27720 non-null  float64
 4   dewpoint          27720 non-null  float64
 5   humidity          27720 non-null  float64
 6   precipitation     27720 non-null  float64
 7   precipprob        27720 non-null  int64  
 8   preciptype        2751 non-null   object 
 9   snow              27702 non-null  float64
 10  snowdepth         27702 non-null  float64
 11  windgust          12223 non-null  float64
 12  windspeed         27720 non-null  float64
 13  winddirection     27720 non-null  float64
 14  sealevelpressure  27702 non-null  float64
 15  cloudcover        27720 non-null  float64
 16  visibility        27702 non-null  float6

In [8]:
demand_forecast_data_2.head(5)

Name       datetime  Temperature  feelslike  dewpoint  humidity  \
0  Pokhara  1/1/2020 0:00         38.0       34.4      29.0     69.76   
1  Pokhara  1/1/2020 1:00         38.0       34.4      29.9     72.37   
2  Pokhara  1/1/2020 2:00         38.0       35.7      30.8     75.06   
3  Pokhara  1/1/2020 3:00         37.1       33.4      29.9     74.97   
4  Pokhara  1/1/2020 4:00         35.1       32.3      29.0     78.41   

   precipitation  precipprob preciptype  snow  ...  visibility  \
0            0.0           0        NaN   0.0  ...         9.9   
1            0.0           0        NaN   0.0  ...         9.9   
2            0.0           0        NaN   0.0  ...         9.9   
3            0.0           0        NaN   0.0  ...         9.9   
4            0.0           0        NaN   0.0  ...         9.9   

   solarradiation  uvindex  severerisk        conditions  Unnamed: 21  \
0             0.0        0         NaN          Overcast          NaN   
1             0.0        0         NaN          Overcast          NaN   
2             0.0        0         NaN          Overcast          NaN   
3             0.0        0         NaN  Partially cloudy          NaN   
4             0.0        0         NaN             Clear          NaN   

   Unnamed: 22  Unnamed: 23  Unnamed: 24  Unnamed: 25  
0          NaN          NaN          NaN          NaN  
1          NaN          NaN          NaN          NaN  
2          NaN          NaN          NaN          NaN  
3          NaN          NaN          NaN          NaN  
4          NaN          NaN          NaN          NaN  

[5 rows x 26 columns]

In [9]:
demand_forecast_data_2.tail()

Name         datetime  Temperature  feelslike  dewpoint  humidity  \
27715  Pokhara  2/28/2023 19:00         56.0       56.0      39.8     54.45   
27716  Pokhara  2/28/2023 20:00         52.1       52.1      41.0     65.83   
27717  Pokhara  2/28/2023 21:00         53.1       53.1      41.0     63.45   
27718  Pokhara  2/28/2023 22:00         55.1       55.1      41.0     58.85   
27719  Pokhara  2/28/2023 23:00         53.1       53.1      41.0     63.45   

       precipitation  precipprob preciptype  snow  ...  visibility  \
27715            0.0           0        NaN   0.0  ...         9.9   
27716            0.0           0        NaN   0.0  ...         9.9   
27717            0.0           0        NaN   0.0  ...         9.9   
27718            0.0           0        NaN   0.0  ...         9.9   
27719            0.0           0        NaN   0.0  ...         9.9   

       solarradiation  uvindex  severerisk  conditions  Unnamed: 21  \
27715             0.0        0        10.0       Clear          NaN   
27716             0.0        0        10.0       Clear          NaN   
27717             0.0        0        10.0       Clear          NaN   
27718             0.0        0        10.0       Clear          NaN   
27719             0.0        0        10.0       Clear          NaN   

       Unnamed: 22  Unnamed: 23  Unnamed: 24  Unnamed: 25  
27715          NaN          NaN          NaN          NaN  
27716          NaN          NaN          NaN          NaN  
27717          NaN          NaN          NaN          NaN  
27718          NaN          NaN          NaN          NaN  
27719          NaN          NaN          NaN          NaN  

[5 rows x 26 columns]

In [10]:
price_forecast_data_path = "Data/Price forecasting/Price Forecasting data upto December 24.csv"
price_forecast_data = pd.read_csv(project_path + price_forecast_data_path)
price_forecast_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35352 entries, 0 to 35351
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Delivery Day      34896 non-null  object 
 1   Hours             34896 non-null  object 
 2   Prices
(EUR/MWh)  34896 non-null  float64
dtypes: float64(1), object(2)
memory usage: 828.7+ KB


In [11]:
price_forecast_data.head()

Delivery Day Hours  Prices\n(EUR/MWh)
0     1/1/2020    H1              26.38
1     1/1/2020    H2              26.50
2     1/1/2020    H3              27.01
3     1/1/2020    H4              27.40
4     1/1/2020    H5              27.99

In [16]:
price_forecast_data.isnull().sum()

Delivery Day         456
Hours                456
Prices\n(EUR/MWh)    456
dtype: int64

In [14]:
df = price_forecast_data.dropna()

In [15]:
df.tail()

Delivery Day Hours  Prices\n(EUR/MWh)
34891   12/24/2023   H20              71.09
34892   12/24/2023   H21              62.01
34893   12/24/2023   H22              53.41
34894   12/24/2023   H23              50.10
34895   12/24/2023   H24              43.23

### H1 to H24 indicates the measuring time, i.e., H1 means 1:00 A.M., H2 means 2:00 A.M., and etc

In [ ]:
price_forecast_data.Hours.unique()

array(['H1', 'H2', 'H3', 'H4', 'H5', 'H6', 'H7', 'H8', 'H9', 'H10', 'H11',
       'H12', 'H13', 'H14', 'H15', 'H16', 'H17', 'H18', 'H19', 'H20',
       'H21', 'H22', 'H23', 'H24', nan], dtype=object)

## Checking for the same dates in the series of above three datasets

In [ ]:
demand_forecast_data_1["datetime"] = pd.to_datetime(demand_forecast_data_1["datetime"], format='%m/%d/%Y %H:%M')
demand_forecast_data_2["datetime"] = pd.to_datetime(demand_forecast_data_2["datetime"], format='%m/%d/%Y %H:%M')

In [ ]:
demand_forecast_data_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27552 entries, 0 to 27551
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   datetime     27552 non-null  datetime64[ns]
 1   Demand (MW)  27552 non-null  float64       
dtypes: datetime64[ns](1), float64(1)
memory usage: 430.6 KB


In [ ]:
## check how many matches are there
merge_inner_demand_data = pd.merge(demand_forecast_data_1, demand_forecast_data_2, on='datetime', how='inner')
num_matches = len(merge_inner_demand_data)
print(f'The number of matches is: {num_matches}')

The number of matches is: 27552


It shows all the demand data has matched datetime with the independent variables dataset

In [ ]:
merge_demand_data = pd.merge(demand_forecast_data_1, demand_forecast_data_2, on='datetime', how='outer')
merge_demand_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 27723 entries, 0 to 27722
Data columns (total 27 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   datetime          27723 non-null  datetime64[ns]
 1   Demand (MW)       27555 non-null  float64       
 2   Name              27720 non-null  object        
 3   Temperature       27720 non-null  float64       
 4   feelslike         27720 non-null  float64       
 5   dewpoint          27720 non-null  float64       
 6   humidity          27720 non-null  float64       
 7   precipitation     27720 non-null  float64       
 8   precipprob        27720 non-null  float64       
 9   preciptype        2751 non-null   object        
 10  snow              27702 non-null  float64       
 11  snowdepth         27702 non-null  float64       
 12  windgust          12223 non-null  float64       
 13  windspeed         27720 non-null  float64       
 14  winddirection     2772

In [ ]:
merge_demand_data.head()

datetime  Demand (MW)     Name  Temperature  feelslike  dewpoint  \
0 2020-01-01 00:00:00        445.8  Pokhara         38.0       34.4      29.0   
1 2020-01-01 01:00:00        424.5  Pokhara         38.0       34.4      29.9   
2 2020-01-01 02:00:00        423.5  Pokhara         38.0       35.7      30.8   
3 2020-01-01 03:00:00        418.8  Pokhara         37.1       33.4      29.9   
4 2020-01-01 04:00:00        414.8  Pokhara         35.1       32.3      29.0   

   humidity  precipitation  precipprob preciptype  ...  visibility  \
0     69.76            0.0         0.0        NaN  ...         9.9   
1     72.37            0.0         0.0        NaN  ...         9.9   
2     75.06            0.0         0.0        NaN  ...         9.9   
3     74.97            0.0         0.0        NaN  ...         9.9   
4     78.41            0.0         0.0        NaN  ...         9.9   

   solarradiation  uvindex  severerisk        conditions  Unnamed: 21  \
0             0.0      0.0         NaN          Overcast          NaN   
1             0.0      0.0         NaN          Overcast          NaN   
2             0.0      0.0         NaN          Overcast          NaN   
3             0.0      0.0         NaN  Partially cloudy          NaN   
4             0.0      0.0         NaN             Clear          NaN   

   Unnamed: 22  Unnamed: 23  Unnamed: 24  Unnamed: 25  
0          NaN          NaN          NaN          NaN  
1          NaN          NaN          NaN          NaN  
2          NaN          NaN          NaN          NaN  
3          NaN          NaN          NaN          NaN  
4          NaN          NaN          NaN          NaN  

[5 rows x 27 columns]

In [ ]:
columns_to_drop = ["Unnamed: 21", "Unnamed: 22", "Unnamed: 23", "Unnamed: 24", "Unnamed: 25"]
merge_demand_data = merge_demand_data.drop(columns=columns_to_drop, axis=1)

In [ ]:
merge_demand_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 27723 entries, 0 to 27722
Data columns (total 22 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   datetime          27723 non-null  datetime64[ns]
 1   Demand (MW)       27555 non-null  float64       
 2   Name              27720 non-null  object        
 3   Temperature       27720 non-null  float64       
 4   feelslike         27720 non-null  float64       
 5   dewpoint          27720 non-null  float64       
 6   humidity          27720 non-null  float64       
 7   precipitation     27720 non-null  float64       
 8   precipprob        27720 non-null  float64       
 9   preciptype        2751 non-null   object        
 10  snow              27702 non-null  float64       
 11  snowdepth         27702 non-null  float64       
 12  windgust          12223 non-null  float64       
 13  windspeed         27720 non-null  float64       
 14  winddirection     2772

In [ ]:
merge_data_path = "/content/drive/MyDrive/Colab Notebooks/Hitachi/Data/Demand Forecasting/"
# merge_demand_data.to_csv(merge_data_path + "merge_demand_data.csv", index=False)

## Working on Price Forecasting DataTime

In [ ]:
price_forecast_data.head()

Delivery Day Hours  Prices\n(EUR/MWh)
0     1/1/2020    H1              26.38
1     1/1/2020    H2              26.50
2     1/1/2020    H3              27.01
3     1/1/2020    H4              27.40
4     1/1/2020    H5              27.99

In [ ]:
price_forecast_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35352 entries, 0 to 35351
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Delivery Day      34896 non-null  object 
 1   Hours             34896 non-null  object 
 2   Prices
(EUR/MWh)  34896 non-null  float64
dtypes: float64(1), object(2)
memory usage: 828.7+ KB


In [ ]:
price_forecast_data = price_forecast_data.dropna()

In [ ]:
price_forecast_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 34896 entries, 0 to 34895
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Delivery Day      34896 non-null  object 
 1   Hours             34896 non-null  object 
 2   Prices
(EUR/MWh)  34896 non-null  float64
dtypes: float64(1), object(2)
memory usage: 1.1+ MB


In [ ]:
# H1 = 1:00
# H2 = 2:00 and so on
# create a dictionary for this
hour_to_time = {}
for i in range(1, 24):
  key = "H" + str(i)
  value = str(i) + ":00"
  hour_to_time[key] = value

hour_to_time["H24"] = "0:00"

print(hour_to_time)

{'H1': '1:00', 'H2': '2:00', 'H3': '3:00', 'H4': '4:00', 'H5': '5:00', 'H6': '6:00', 'H7': '7:00', 'H8': '8:00', 'H9': '9:00', 'H10': '10:00', 'H11': '11:00', 'H12': '12:00', 'H13': '13:00', 'H14': '14:00', 'H15': '15:00', 'H16': '16:00', 'H17': '17:00', 'H18': '18:00', 'H19': '19:00', 'H20': '20:00', 'H21': '21:00', 'H22': '22:00', 'H23': '23:00', 'H24': '0:00'}


In [ ]:
datetime_column = []
for index, row in price_forecast_data.iterrows():
    delivery_day = row['Delivery Day']
    hours = row['Hours']
    time = hour_to_time[hours]
    # print(delivery_day + " "+ time)
    datetime_column.append(delivery_day + " "+ time)

In [ ]:
price_forecast_data["datetime"] = datetime_column

In [ ]:
price_forecast_data.head()

Delivery Day Hours  Prices\n(EUR/MWh)       datetime
0     1/1/2020    H1              26.38  1/1/2020 1:00
1     1/1/2020    H2              26.50  1/1/2020 2:00
2     1/1/2020    H3              27.01  1/1/2020 3:00
3     1/1/2020    H4              27.40  1/1/2020 4:00
4     1/1/2020    H5              27.99  1/1/2020 5:00

In [ ]:
price_forecast_path = "/content/drive/MyDrive/Colab Notebooks/Hitachi/Data/Price forecasting/"
# price_forecast_data.to_csv(price_forecast_path + "updated_price_forecast.csv", index=False)

In [ ]:
# change the datetime object type to datetime type
price_forecast_data["datetime"] = pd.to_datetime(price_forecast_data["datetime"], format='%m/%d/%Y %H:%M')
# check how many matches are there in datetime between demand and price data
merge_inner_data = pd.merge(price_forecast_data, merge_demand_data, on='datetime', how='inner')
num_matches = len(merge_inner_data)
print(f'The number of matches is: {num_matches}')

The number of matches is: 27723


Since, the demand and price have 27723 matches. Most probably, we will be using these data for the price forecasting

In [ ]:
merged_whole_data = pd.merge(price_forecast_data, merge_demand_data, on='datetime', how='outer')
merged_whole_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 34899 entries, 0 to 34898
Data columns (total 25 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Delivery Day      34899 non-null  object        
 1   Hours             34899 non-null  object        
 2   Prices
(EUR/MWh)  34899 non-null  float64       
 3   datetime          34899 non-null  datetime64[ns]
 4   Demand (MW)       27555 non-null  float64       
 5   Name              27720 non-null  object        
 6   Temperature       27720 non-null  float64       
 7   feelslike         27720 non-null  float64       
 8   dewpoint          27720 non-null  float64       
 9   humidity          27720 non-null  float64       
 10  precipitation     27720 non-null  float64       
 11  precipprob        27720 non-null  float64       
 12  preciptype        2751 non-null   object        
 13  snow              27702 non-null  float64       
 14  snowdepth         2770

In [ ]:
merge_whole_data_path = "/content/drive/MyDrive/Colab Notebooks/Hitachi/Data/"
# merged_whole_data.to_csv(merge_whole_data_path + "merged_whole_data.csv", index=False)

In [ ]:
merged_whole_data.head()

Delivery Day Hours  Prices\n(EUR/MWh)            datetime  Demand (MW)  \
0     1/1/2020    H1              26.38 2020-01-01 01:00:00        424.5   
1     1/1/2020    H2              26.50 2020-01-01 02:00:00        423.5   
2     1/1/2020    H3              27.01 2020-01-01 03:00:00        418.8   
3     1/1/2020    H4              27.40 2020-01-01 04:00:00        414.8   
4     1/1/2020    H5              27.99 2020-01-01 05:00:00        432.1   

      Name  Temperature  feelslike  dewpoint  humidity  ...  windgust  \
0  Pokhara         38.0       34.4      29.9     72.37  ...       NaN   
1  Pokhara         38.0       35.7      30.8     75.06  ...       NaN   
2  Pokhara         37.1       33.4      29.9     74.97  ...       NaN   
3  Pokhara         35.1       32.3      29.0     78.41  ...       NaN   
4  Pokhara         34.1       29.8      28.1     78.64  ...       NaN   

   windspeed winddirection  sealevelpressure  cloudcover  visibility  \
0        4.7         230.0            1015.3       100.0         9.9   
1        3.4         230.0            1015.4       100.0         9.9   
2        4.7         220.0            1015.3        88.9         9.9   
3        3.4         240.0            1015.2         0.0         9.9   
4        4.7         230.0            1015.4         0.0         9.9   

   solarradiation  uvindex  severerisk        conditions  
0             0.0      0.0         NaN          Overcast  
1             0.0      0.0         NaN          Overcast  
2             0.0      0.0         NaN  Partially cloudy  
3             0.0      0.0         NaN             Clear  
4             0.0      0.0         NaN             Clear  

[5 rows x 25 columns]

In [ ]:
merged_whole_data["Name"].unique()

array(['Pokhara', nan], dtype=object)

# Conclusion

* 27552 entries in demand_forecast_data_1 with 2 columns: datetime and demand   
* 27720 entries in demand_forecast_data_2 with 26 columns
* 35352 entries in price_forecast_data with 3 columns: datetime(delivery day, hours) and Prices

## Basic Data Cleaning

* 5 columns in demand_forecsat_data_2 (Unnamed: 21 ... Unnamed : 25) NULL DATA removed
* There are 34896 NOT NULL ROWS in the price_forecast dataset. Hence all the NULL Rows were dropped.

## Data Transformation
* Data in price_forecast has Delivery Day, Hours columns in the format of (1/1/2020, H1)	which was converted into the datetime format as 1/1/2020 1:00

## New Dataset formed by merging
* merge_demand_data.csv is created by performing outer join on datetime column for the demand_forecast_data_1 and demand_forecast_data_2
* updated_price_forecast.csv is created by performing the transformation of Deliver Day and Hours into datetime as mentioned in Data Transformation section after the drop of all the NULL ROWS.
* merged_whole_data is created by performing the outer join on datetime column for the merge_demand_data and the updated_price_forecast data.
